stl/inc/span

```diff
@@ -0,0 +1,732 @@
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        return const_iterator(data(), size(), size());
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+        return const_iterator(data() + size());
+#endif // _ITERATOR_DEBUG_LEVEL
+    }
+
+#ifdef __cpp_lib_concepts
+    _NODISCARD friend constexpr const_iterator begin(span _Right) noexcept {
+        // non-member overload to model the exposition-only forwarding-range concept
+        return _Right.begin();
+    }
+    _NODISCARD friend constexpr const_iterator end(span _Right) noexcept {
+        // non-member overload to model the exposition-only forwarding-range concept
+        return _Right.end();
+    }
+#endif // __cpp_lib_concepts
+
+    _NODISCARD constexpr const_iterator cbegin() const noexcept {
+        return begin();
+    }
+
+    _NODISCARD constexpr const_iterator cend() const noexcept {
+        return end();
+    }
+
+    _NODISCARD constexpr reverse_iterator rbegin() noexcept {
+        return reverse_iterator{end()};
+    }
+
+    _NODISCARD constexpr reverse_iterator rend() noexcept {
+        return reverse_iterator{begin()};
+    }
+
+    _NODISCARD constexpr const_reverse_iterator rbegin() const noexcept {
+        return const_reverse_iterator{end()};
+    }
+
+    _NODISCARD constexpr const_reverse_iterator rend() const noexcept {
+        return const_reverse_iterator{begin()};
+    }
+
+    _NODISCARD constexpr const_reverse_iterator crbegin() const noexcept {
+        return rbegin();
+    }
+
+    _NODISCARD constexpr const_reverse_iterator crend() const noexcept {
+        return rend();
+    }
+
+    _NODISCARD constexpr pointer _Unchecked_begin() noexcept {
+        return data();
+    }
+
+    _NODISCARD constexpr const_pointer _Unchecked_begin() const noexcept {
+        return data();
+    }
+
+    _NODISCARD constexpr pointer _Unchecked_end() noexcept {
+        return data() + size();
+    }
+
+    _NODISCARD constexpr const_pointer _Unchecked_end() const noexcept {
+        return data() + size();
+    }
+
+    constexpr void swap(span& _Other) noexcept {
+        _Mydata.swap(_Other._Mydata);
+    }
+
+    // [span.objectrep] Views of object representation
+    _NODISCARD span<const byte, _Extent == dynamic_extent ? dynamic_extent : sizeof(element_type) * _Extent>
+        _As_bytes() const noexcept {
+        return {reinterpret_cast<const byte*>(data()), size_bytes()};
+    }
+
+    _NODISCARD span<byte, _Extent == dynamic_extent ? dynamic_extent : sizeof(element_type) * _Extent>
+        _As_writeable_bytes() const noexcept {
+        return {reinterpret_cast<byte*>(data()), size_bytes()};
+    }
+
+private:
+    _Span_storage<pointer, _Extent> _Mydata;
+};
+
+// [span.tuple] Tuple interface
+template <class _Ty, size_t _Extent>
+struct tuple_size<span<_Ty, _Extent>> : public integral_constant<size_t, _Extent> {};
+
+template <class _Ty>
+struct tuple_size<span<_Ty, dynamic_extent>>;

```

Comment:
It is unclear to me what the Standard says here. Making this incomplete seems incorrect. [span.tuple]/1 seems to imply that we should `static_assert`.

Reply:
Yeah I want sure what to do about that and went with what libcxx does. We could assert in the first definition and remove that?

Reply:
I think that `static_assert`ing in the first definition is correct, yes.

Regarding your comment about libcxx - please ensure that you're developing your implementation independently. (We added a checklist item to the PR template after you created this PR.) If it is instead *derived* from libcxx sources, we need to know, so we can ensure that the licenses and notices are properly respected.

Reply:
I believe the implementation is quite distinct from that of libcxx. 
Nevertheless for hairy details I tend to look around how others solve the problem. Either in STL of libcxx